# Pandas: 
### - Calling APIs using Python

<strong> <h2> Run The Following Code </h2> </strong>

Hit SHIFT + ENTER when your cusor is inside the cell of code.<br/>

> Import the packages

In [1]:
import pandas as pd
import requests

## Calling API Example 1

> 1. Open this url in your browser: https://data.gov.sg/dataset/list-of-government-markets-hawker-centres
> 2. Click on the "Data API" button on top rigth corner to see resource_id for this dataset
> 3. Use **requests** package to call this API and get all first 5 rows of data

In [2]:
# Option A - Manually construct the request URL
url_full = 'https://data.gov.sg/api/action/datastore_search?resource_id=8f6bba57-19fc-4f36-8dcf-c0bda382364d&limit=5'

response = requests.get(url_full)

In [3]:
# Option B - (Recommneded) passing a dictionary to .get() method in requests to construct the request url
url_base = 'https://data.gov.sg/api/action/datastore_search'

parameters = {
    'resource_id' : '8f6bba57-19fc-4f36-8dcf-c0bda382364d',
    'limit': '5'
}
response = requests.get(url_base, params=parameters)

In [4]:
# Check the url that sent to the API server
response.url

'https://data.gov.sg/api/action/datastore_search?resource_id=8f6bba57-19fc-4f36-8dcf-c0bda382364d&limit=5'

In [5]:
# Check the status of the request
response.status_code

200

In [6]:
# View the json text returned by the API server
response_dict = response.json()
response_dict

{'help': 'https://data.gov.sg/api/3/action/help_show?name=datastore_search',
 'success': True,
 'result': {'resource_id': '8f6bba57-19fc-4f36-8dcf-c0bda382364d',
  'fields': [{'type': 'int4', 'id': '_id'},
   {'type': 'text', 'id': 'name_of_centre'},
   {'type': 'text', 'id': 'location_of_centre'},
   {'type': 'text', 'id': 'type_of_centre'},
   {'type': 'text', 'id': 'owner'},
   {'type': 'numeric', 'id': 'no_of_stalls'},
   {'type': 'numeric', 'id': 'no_of_cooked_food_stalls'},
   {'type': 'numeric', 'id': 'no_of_mkt_produce_stalls'}],
  'records': [{'location_of_centre': '2, Adam Road, S(289876)',
    'no_of_cooked_food_stalls': '32',
    'no_of_mkt_produce_stalls': '0',
    'name_of_centre': 'Adam Road Food Centre',
    'type_of_centre': 'HC',
    'no_of_stalls': '32',
    'owner': 'Government',
    '_id': 1},
   {'location_of_centre': 'National Development Building, Annex B, Telok Ayer Street, S(069111)',
    'no_of_cooked_food_stalls': '134',
    'no_of_mkt_produce_stalls': '1',


> The json text above is a nested-dictionary <br>
> Picture below shows the structure in a visual form

> ![](hawker_structure.png)

In [7]:
# Traverse the dictionary to the level where the main records are located
response_dict['result']['total']

107

In [8]:
# Traverse the dictionary to the level where the main records are located
records = response_dict['result']['records']
records

[{'location_of_centre': '2, Adam Road, S(289876)',
  'no_of_cooked_food_stalls': '32',
  'no_of_mkt_produce_stalls': '0',
  'name_of_centre': 'Adam Road Food Centre',
  'type_of_centre': 'HC',
  'no_of_stalls': '32',
  'owner': 'Government',
  '_id': 1},
 {'location_of_centre': 'National Development Building, Annex B, Telok Ayer Street, S(069111)',
  'no_of_cooked_food_stalls': '134',
  'no_of_mkt_produce_stalls': '1',
  'name_of_centre': 'Amoy Street Food Centre',
  'type_of_centre': 'HC',
  'no_of_stalls': '135',
  'owner': 'Government',
  '_id': 2},
 {'location_of_centre': '1, Bedok Road, S(469572)',
  'no_of_cooked_food_stalls': '32',
  'no_of_mkt_produce_stalls': '0',
  'name_of_centre': 'Bedok Food Centre',
  'type_of_centre': 'HC',
  'no_of_stalls': '32',
  'owner': 'Government',
  '_id': 3},
 {'location_of_centre': '38A, Beo Crescent, S(169982)',
  'no_of_cooked_food_stalls': '32',
  'no_of_mkt_produce_stalls': '62',
  'name_of_centre': 'Beo Crescent Market',
  'type_of_centre'

In [9]:
# Check the number of records
len(records)

5

In [10]:
# Use .json_normalize from Pandas to convert the dictionary into a DataFrame
pd.json_normalize(records)

location_of_centre no_of_cooked_food_stalls  \
0                            2, Adam Road, S(289876)                       32   
1  National Development Building, Annex B, Telok ...                      134   
2                           1, Bedok Road, S(469572)                       32   
3                       38A, Beo Crescent, S(169982)                       32   
4                        166, Jalan Besar, S(208877)                       66   

  no_of_mkt_produce_stalls           name_of_centre type_of_centre  \
0                        0    Adam Road Food Centre             HC   
1                        1  Amoy Street Food Centre             HC   
2                        0        Bedok Food Centre             HC   
3                       62      Beo Crescent Market            MHC   
4                        0       Berseh Food Centre             HC   

  no_of_stalls       owner  _id  
0           32  Government    1  
1          135  Government    2  
2           32  Government    3  
3           94  Government    4  
4           66  Government    5

In [11]:
# Use .json_normalize from Pandas to convert the dictionary into a DataFrame 
# Store into a variable called "df_hawker"
df_hawker = pd.json_normalize(records)

## Calling API Example 2: with Parameters

> 1. Open this url in your browser: https://www.onemap.gov.sg/docs/#onemap-rest-apis
> 2. Understand the parameters (required and optional) of the "search" endpoint from OneMap API
> 3. Use **requests** package to call this API and get all first 5 rows of data

> ![apiexp](onemap_search.png)

In [12]:
# Try to call the API (Case-sensitive Parameters not match)

url_base = 'https://developers.onemap.sg/commonapi/search'

parameters = {
    'searchval' : 'Adam Road Food Centre',
    'returngeom': 'Y',
    'getAddrDetails': 'Y'
}
response = requests.get(url_base, params=parameters)
response.status_code
# Outcome is 400 status code

400

In [13]:
# Try to call the API (Required Parameters not provided)

url_base = 'https://developers.onemap.sg/commonapi/search'

parameters = {
    'searchval' : 'Adam Road Food Centre',
}
response = requests.get(url_base, params=parameters)
response.status_code
# Outcome is 400 series status code

400

In [14]:
# Try to call the API (Correctly)

url_base = 'https://developers.onemap.sg/commonapi/search'

parameters = {
    'searchVal' : 'Adam Road Food Centre',
    'returnGeom': 'Y',
    'getAddrDetails': 'Y'
}
response = requests.get(url_base, params=parameters)
response.status_code

200

In [15]:
# View the json text returned by the server
response.json()

{'found': 1,
 'totalNumPages': 1,
 'pageNum': 1,
 'results': [{'SEARCHVAL': 'ADAM ROAD FOOD CENTRE',
   'BLK_NO': '2',
   'ROAD_NAME': 'ADAM ROAD',
   'BUILDING': 'ADAM ROAD FOOD CENTRE',
   'ADDRESS': '2 ADAM ROAD ADAM ROAD FOOD CENTRE SINGAPORE 289876',
   'POSTAL': '289876',
   'X': '25870.3018411137',
   'Y': '34035.8644314632',
   'LATITUDE': '1.3240827139625',
   'LONGITUDE': '103.814182099841',
   'LONGTITUDE': '103.814182099841'}]}

In [16]:
# Use .json_normalize from Pandas to convert the dictionary into a DataFrame 
pd.json_normalize(response.json()['results'])

SEARCHVAL BLK_NO  ROAD_NAME               BUILDING  \
0  ADAM ROAD FOOD CENTRE      2  ADAM ROAD  ADAM ROAD FOOD CENTRE   

                                             ADDRESS  POSTAL  \
0  2 ADAM ROAD ADAM ROAD FOOD CENTRE SINGAPORE 28...  289876   

                  X                 Y         LATITUDE         LONGITUDE  \
0  25870.3018411137  34035.8644314632  1.3240827139625  103.814182099841   

         LONGTITUDE  
0  103.814182099841

> 1. Loop through the **df_hawker** and find the geoinfo of each hawker from OneMap's SEARCH endpoint
> 2. Store all the geoinfo as a new DataFrame

In [17]:
len(df_hawker)

5

In [18]:
df_hawker_geoinfo = []

for row_index, row in df_hawker.iterrows():
    # Create a dictionary of the parameters
    parameters = {
        'searchVal' : row['name_of_centre'],
        'returnGeom': 'Y',
        'getAddrDetails': 'Y'
    }

    # Send the request via .get() method
    response = requests.get(url_base, params=parameters)

    # Check if the request is successful
    if response.status_code == 200:
        response_dict = response.json()
        # Check if there is at least 1 record
        if len(response_dict['results']) > 0:
            df_temp = pd.json_normalize(response_dict['results'])
            df_hawker_geoinfo.append(df_temp)
    else:
        print('The request failed and returned status code:' + str(response.status_code))

In [19]:
# Concatenate the records stored in multiple DataFrames into a single DataFrame
df_hawker_geoinfo = pd.concat(df_hawker_geoinfo, axis=0, ignore_index=True)
df_hawker_geoinfo

SEARCHVAL BLK_NO     ROAD_NAME                 BUILDING  \
0    ADAM ROAD FOOD CENTRE      2     ADAM ROAD    ADAM ROAD FOOD CENTRE   
1  AMOY STREET FOOD CENTRE      7  MAXWELL ROAD  AMOY STREET FOOD CENTRE   
2        BEDOK FOOD CENTRE      1    BEDOK ROAD        BEDOK FOOD CENTRE   
3      BEO CRESCENT MARKET    38A  BEO CRESCENT      BEO CRESCENT MARKET   
4       BERSEH FOOD CENTRE    166   JALAN BESAR       BERSEH FOOD CENTRE   

                                             ADDRESS  POSTAL  \
0  2 ADAM ROAD ADAM ROAD FOOD CENTRE SINGAPORE 28...  289876   
1  7 MAXWELL ROAD AMOY STREET FOOD CENTRE SINGAPO...  069111   
2    1 BEDOK ROAD BEDOK FOOD CENTRE SINGAPORE 469572  469572   
3  38A BEO CRESCENT BEO CRESCENT MARKET SINGAPORE...  169982   
4  166 JALAN BESAR BERSEH FOOD CENTRE SINGAPORE 2...  208877   

                  X                 Y          LATITUDE         LONGITUDE  \
0  25870.3018411137  34035.8644314632   1.3240827139625  103.814182099841   
1   29483.955767219  29088.4261356071   1.2793398636571  103.846652482254   
2  41595.4345044643   33623.132626934  1.32034716834128  103.955480570427   
3  27336.1809281698   30137.891893005  1.28883089150258  103.827353892228   
4  30623.1448179465  32184.9996254593  1.30734410946371  103.856888783769   

         LONGTITUDE  
0  103.814182099841  
1  103.846652482254  
2  103.955480570427  
3  103.827353892228  
4  103.856888783769

In [20]:
len(df_hawker_geoinfo)

5

# !! Your Turn !!

## Replace the code marked as <..> COMPLETELY with your own.
- The placeholder <..> is meant to be guidance for your answer.
- It should not be restricting your solutions for the questions.
- Feel free to add more lines or use less lines,
- One placeholder does not mean that you can only insert one line of codes.

# Question 1
> - Continue from the **df_hawker** and **df_hawker_geoinfo**, produce a new DataFrame **df_hawker_center**, <br>
> where each row is a unique hawker center with all the columns both **df_hawker** and **df_hawker_geoinfo**
> - The four columns "X", "Y", "LONGTITUDE", and "_id" must be excluded in the **df_hawker_center**
> - Change all the column names into lower case

In [21]:
# Common column to merge on is the 'name of center' from the df_hawker dataframe and the SEARCHVAL columns
# from the df_hawker_geoinfo dataframe
# But first, let's we normalise the case of one of the commmon columns to upperer case
df_hawker['name_of_centre'] = df_hawker['name_of_centre'].apply(lambda x: x.upper())

In [22]:
#  Merge the two dataframes
df_hawker_center = df_hawker.merge(df_hawker_geoinfo, how='left', left_on='name_of_centre', right_on='SEARCHVAL')

In [23]:
# Drop columns X, Y and longtitude

df_hawker_center.drop(['X', 'Y', 'LONGTITUDE', '_id'], axis=1, inplace=True)

In [24]:
# Change all column names to lower case
df_hawker_center = df_hawker_center.rename(columns=str.lower)

df_hawker_center

location_of_centre no_of_cooked_food_stalls  \
0                            2, Adam Road, S(289876)                       32   
1  National Development Building, Annex B, Telok ...                      134   
2                           1, Bedok Road, S(469572)                       32   
3                       38A, Beo Crescent, S(169982)                       32   
4                        166, Jalan Besar, S(208877)                       66   

  no_of_mkt_produce_stalls           name_of_centre type_of_centre  \
0                        0    ADAM ROAD FOOD CENTRE             HC   
1                        1  AMOY STREET FOOD CENTRE             HC   
2                        0        BEDOK FOOD CENTRE             HC   
3                       62      BEO CRESCENT MARKET            MHC   
4                        0       BERSEH FOOD CENTRE             HC   

  no_of_stalls       owner                searchval blk_no     road_name  \
0           32  Government    ADAM ROAD FOOD CENTRE      2     ADAM ROAD   
1          135  Government  AMOY STREET FOOD CENTRE      7  MAXWELL ROAD   
2           32  Government        BEDOK FOOD CENTRE      1    BEDOK ROAD   
3           94  Government      BEO CRESCENT MARKET    38A  BEO CRESCENT   
4           66  Government       BERSEH FOOD CENTRE    166   JALAN BESAR   

                  building                                            address  \
0    ADAM ROAD FOOD CENTRE  2 ADAM ROAD ADAM ROAD FOOD CENTRE SINGAPORE 28...   
1  AMOY STREET FOOD CENTRE  7 MAXWELL ROAD AMOY STREET FOOD CENTRE SINGAPO...   
2        BEDOK FOOD CENTRE    1 BEDOK ROAD BEDOK FOOD CENTRE SINGAPORE 469572   
3      BEO CRESCENT MARKET  38A BEO CRESCENT BEO CRESCENT MARKET SINGAPORE...   
4       BERSEH FOOD CENTRE  166 JALAN BESAR BERSEH FOOD CENTRE SINGAPORE 2...   

   postal          latitude         longitude  
0  289876   1.3240827139625  103.814182099841  
1  069111   1.2793398636571  103.846652482254  
2  469572  1.32034716834128  103.955480570427  
3  169982  1.28883089150258  103.827353892228  
4  208877  1.30734410946371  103.856888783769

# Question 2


> Part A) <br>
> Store the first 200 records from the Excel file located at **data > 0_raw_data > hdb_postal_to_stations.xlsx** into a DataFrame, called **df_mrt_subset**

In [25]:
df_mrt_subset = pd.read_excel('data\\0_raw_data\hdb_postal_to_stations.xlsx').head(200)

In [26]:
len(df_mrt_subset)

200

In [27]:
df_mrt_subset.head()

postal station_type  station_name  station_nearest_distance
0  560174.0          MRT  Yio Chu Kang               1098.719927
1  560541.0          MRT    Ang Mo Kio                806.155872
2  560163.0          MRT  Yio Chu Kang               1179.564586
3  560446.0          MRT    Ang Mo Kio                688.604642
4  560557.0          MRT    Ang Mo Kio                929.151815

In [28]:
df_mrt_subset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 4 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   postal                    200 non-null    object 
 1   station_type              200 non-null    object 
 2   station_name              200 non-null    object 
 3   station_nearest_distance  200 non-null    float64
dtypes: float64(1), object(3)
memory usage: 6.4+ KB


In [29]:
df_mrt_subset['postal']=df_mrt_subset['postal'].astype('float')
df_mrt_subset['postal']=df_mrt_subset['postal'].astype('int')

In [30]:
df_mrt_subset.head()

postal station_type  station_name  station_nearest_distance
0  560174          MRT  Yio Chu Kang               1098.719927
1  560541          MRT    Ang Mo Kio                806.155872
2  560163          MRT  Yio Chu Kang               1179.564586
3  560446          MRT    Ang Mo Kio                688.604642
4  560557          MRT    Ang Mo Kio                929.151815

> Part B) <br>
> - Retrieve the geoinfo using OneMap's API Endpoint 'https://developers.onemap.sg/commonapi/search', <br>
> using the postal codes of the MRT stations
> - Store the geoinfo for the MRT stations' postal code into DataFrame **df_mrt_geoinfo** <br>
> 💡Hint: You will need to the remove the decimal point at the end of each postal code

In [31]:
df_mrt_geoinfo = []

for row_index, row in df_mrt_subset.iterrows():
    # Create a dictionary of the parameters
    parameters = {
        'searchVal' : row['postal'],
        'returnGeom': 'Y',
        'getAddrDetails': 'Y'
    }

    # Send the request via .get() method
    response = requests.get(url_base, params=parameters)

    # Check if the request is successful
    if response.status_code == 200:
        response_dict = response.json()
        # Check if there is at least 1 record
        if len(response_dict['results']) > 0:
            df_temp = pd.json_normalize(response_dict['results'])
            df_mrt_geoinfo.append(df_temp)
    else:
        print('The request failed and returned status code:' + str(response.status_code))

In [32]:
# Concatenate the records stored in multiple DataFrames into a single DataFrame
df_mrt_geoinfo = pd.concat(df_mrt_geoinfo, axis=0, ignore_index=True)
df_mrt_geoinfo

SEARCHVAL BLK_NO  \
0                                 KEBUN BARU LINK 1    174   
1                                   CHENG SAN GREEN    541   
2                                    MY FIRST SKOOL    541   
3          163 ANG MO KIO AVENUE 4 SINGAPORE 560163    163   
4                                        FULL MARKS    163   
..                                              ...    ...   
236                                     SPRING VIEW   291B   
237                                     NATURE VIEW   288D   
238                         SKY VISTA @ BUKIT BATOK   113C   
239  186 BUKIT BATOK WEST AVENUE 6 SINGAPORE 650186    186   
240  386 BUKIT BATOK WEST AVENUE 5 SINGAPORE 650386    386   

                     ROAD_NAME                 BUILDING  \
0          ANG MO KIO AVENUE 4        KEBUN BARU LINK 1   
1         ANG MO KIO AVENUE 10          CHENG SAN GREEN   
2         ANG MO KIO AVENUE 10           MY FIRST SKOOL   
3          ANG MO KIO AVENUE 4                      NIL   
4          ANG MO KIO AVENUE 4               FULL MARKS   
..                         ...                      ...   
236      BUKIT BATOK STREET 24              SPRING VIEW   
237      BUKIT BATOK STREET 25              NATURE VIEW   
238  BUKIT BATOK WEST AVENUE 6  SKY VISTA @ BUKIT BATOK   
239  BUKIT BATOK WEST AVENUE 6                      NIL   
240  BUKIT BATOK WEST AVENUE 5                      NIL   

                                               ADDRESS  POSTAL  \
0    174 ANG MO KIO AVENUE 4 KEBUN BARU LINK 1 SING...  560174   
1    541 ANG MO KIO AVENUE 10 CHENG SAN GREEN SINGA...  560541   
2    541 ANG MO KIO AVENUE 10 MY FIRST SKOOL SINGAP...  560541   
3             163 ANG MO KIO AVENUE 4 SINGAPORE 560163  560163   
4    163 ANG MO KIO AVENUE 4 FULL MARKS SINGAPORE 5...  560163   
..                                                 ...     ...   
236  291B BUKIT BATOK STREET 24 SPRING VIEW SINGAPO...  651291   
237  288D BUKIT BATOK STREET 25 NATURE VIEW SINGAPO...  653288   
238  113C BUKIT BATOK WEST AVENUE 6 SKY VISTA @ BUK...  653113   
239     186 BUKIT BATOK WEST AVENUE 6 SINGAPORE 650186  650186   
240     386 BUKIT BATOK WEST AVENUE 5 SINGAPORE 650386  650386   

                    X                 Y          LATITUDE         LONGITUDE  \
0    28478.5794445509  39676.8076131288  1.37509746867904   103.83761896123   
1     30482.026548801  39546.8847144619  1.37392239168826  103.855621371068   
2     30482.026488265  39546.8841999061  1.37392238703482  103.855621370524   
3     28540.623786686   39505.534986278  1.37354853919927  103.838176471398   
4     28539.787784976  39505.8939948324  1.37355178595963  103.838168959329   
..                ...               ...               ...               ...   
236  19379.9538016999  36145.1296477765  1.34315699982096  103.755862270481   
237  19758.0990471318  36437.7046699263  1.34580304817683  103.759260051639   
238  18209.7822006959  36831.8956999574  1.34936750240674  103.745347358824   
239  18069.3259528903  36523.7574556731  1.34658076680317  103.744085374214   
240   19012.082054765  38443.1579966731  1.36393942184816  103.752556035691   

           LONGTITUDE  
0     103.83761896123  
1    103.855621371068  
2    103.855621370524  
3    103.838176471398  
4    103.838168959329  
..                ...  
236  103.755862270481  
237  103.759260051639  
238  103.745347358824  
239  103.744085374214  
240  103.752556035691  

[241 rows x 11 columns]

# Question 3

> Part A) <br>
> - Download all the records for HDB carpark info from the API https://data.gov.sg/dataset/hdb-carpark-information
> - Store the records into a DataFrame, called **df_carpark**

In [33]:
url_base = 'https://data.gov.sg/api/action/datastore_search'

parameters = {
    'resource_id' : '139a3035-e624-4f56-b63f-89ae28d4ae4c',
}
response = requests.get(url_base, params=parameters)

In [34]:
# View the json text returned by the API server
response_dict = response.json()
response_dict

{'help': 'https://data.gov.sg/api/3/action/help_show?name=datastore_search',
 'success': True,
 'result': {'resource_id': '139a3035-e624-4f56-b63f-89ae28d4ae4c',
  'fields': [{'type': 'int4', 'id': '_id'},
   {'type': 'text', 'id': 'car_park_no'},
   {'type': 'text', 'id': 'address'},
   {'type': 'numeric', 'id': 'x_coord'},
   {'type': 'numeric', 'id': 'y_coord'},
   {'type': 'text', 'id': 'car_park_type'},
   {'type': 'text', 'id': 'type_of_parking_system'},
   {'type': 'text', 'id': 'short_term_parking'},
   {'type': 'text', 'id': 'free_parking'},
   {'type': 'text', 'id': 'night_parking'},
   {'type': 'numeric', 'id': 'car_park_decks'},
   {'type': 'numeric', 'id': 'gantry_height'},
   {'type': 'text', 'id': 'car_park_basement'}],
  'records': [{'short_term_parking': 'WHOLE DAY',
    'car_park_type': 'BASEMENT CAR PARK',
    'y_coord': '31490.4942',
    'x_coord': '30314.7936',
    'free_parking': 'NO',
    'gantry_height': '1.80',
    'car_park_basement': 'Y',
    'night_parking':

In [35]:

# Traverse the dictionary to the level where the main records are located
response_dict['result']['total']

2183

In [36]:
# Traverse the dictionary to the level where the main records are located
records = response_dict['result']['records']

# Check the number of records
len(records)



100

In [37]:
# Use .json_normalize from Pandas to convert the dictionary into a DataFrame
df_carpark = pd.json_normalize(records)

df_carpark

short_term_parking          car_park_type     y_coord     x_coord  \
0           WHOLE DAY      BASEMENT CAR PARK  31490.4942  30314.7936   
1           WHOLE DAY  MULTI-STOREY CAR PARK  33695.5198  33758.4143   
2           WHOLE DAY       SURFACE CAR PARK  34500.3599  29257.7203   
3             7AM-7PM       SURFACE CAR PARK  39012.6664  28185.4359   
4                  NO       SURFACE CAR PARK  38684.1754  29482.0290   
..                ...                    ...         ...         ...   
95          WHOLE DAY       SURFACE CAR PARK  40303.0500  29136.4400   
96                 NO       SURFACE CAR PARK  28847.7781  28733.8679   
97          WHOLE DAY       SURFACE CAR PARK  40091.5441  28835.3369   
98          WHOLE DAY       SURFACE CAR PARK  40069.8164  29166.9700   
99          WHOLE DAY       SURFACE CAR PARK  39798.9610  28597.1669   

               free_parking gantry_height car_park_basement night_parking  \
0                        NO          1.80                 Y           YES   
1   SUN & PH FR 7AM-10.30PM          2.10                 N           YES   
2   SUN & PH FR 7AM-10.30PM          0.00                 N           YES   
3                        NO          0.00                 N            NO   
4                        NO          0.00                 N            NO   
..                      ...           ...               ...           ...   
95  SUN & PH FR 7AM-10.30PM          5.40                 N           YES   
96                       NO          4.50                 N            NO   
97  SUN & PH FR 7AM-10.30PM          0.00                 N           YES   
98  SUN & PH FR 7AM-10.30PM          0.00                 N           YES   
99  SUN & PH FR 7AM-10.30PM          4.50                 N           YES   

                                        address car_park_decks  _id  \
0   BLK 270/271 ALBERT CENTRE BASEMENT CAR PARK              1    1   
1                     BLK 98A ALJUNIED CRESCENT              5    2   
2                           BLK 101 JALAN DUSUN              0    3   
3                BLOCK 253 ANG MO KIO STREET 21              0    4   
4                  BLK 302/348 ANG MO KIO ST 31              0    5   
..                                          ...            ...  ...   
95                 BLK 633/640 ANG MO KIO ST 61              0   96   
96                         BLK 1/7 EVERTON PARK              0  612   
97                 BLK 641/645 ANG MO KIO ST 61              0   97   
98                 BLK 646/649 ANG MO KIO ST 61              0   98   
99                 BLK 177/182 ANG MO KIO AVE 4              0   99   

   car_park_no type_of_parking_system  
0          ACB     ELECTRONIC PARKING  
1          ACM     ELECTRONIC PARKING  
2          AH1     ELECTRONIC PARKING  
3         AK19         COUPON PARKING  
4         AK31         COUPON PARKING  
..         ...                    ...  
95         A72     ELECTRONIC PARKING  
96         EPL     ELECTRONIC PARKING  
97         A73     ELECTRONIC PARKING  
98         A74     ELECTRONIC PARKING  
99         A75     ELECTRONIC PARKING  

[100 rows x 13 columns]

> Part B)** - Challenging Question [Optional] <br>
> - Check the number of records in **df_carpark**
> - You will realize the total records on data.gov.sg is 2,183 records, but **df_carpark**only has 100 records
> - This is because it's common for API return a "page" of results per call. In this case, the API returns 100 records per page.
> - If you check *response.json()['result']['_links']*, you will see there are two links, the "next" link shows how to retrieve the next 100 records
> - Find a way to download all 2,183 records and store into the dataframe **df_carparks**

> hint: you can use anything that you have learnt so far to achieve this

In [38]:
response.json()['result']['_links']

{'start': '/api/action/datastore_search?resource_id=139a3035-e624-4f56-b63f-89ae28d4ae4c',
 'next': '/api/action/datastore_search?offset=100&resource_id=139a3035-e624-4f56-b63f-89ae28d4ae4c'}

In [39]:
url_base = 'https://data.gov.sg/api/action/datastore_search'


# Use enter a large number in the limit parameter to retrieve all the records
parameters = {
    'resource_id' : '139a3035-e624-4f56-b63f-89ae28d4ae4c',
    'limit' : 1000000
}
response = requests.get(url_base, params=parameters)

response_dict = response.json()

records = response_dict['result']['records']

len(records)

2183

In [40]:
# Use .json_normalize from Pandas to convert the dictionary into a DataFrame
df_carparks = pd.json_normalize(records)

len(df_carparks)

2183

## Method 2:  Looping through every 100 records

In [41]:
#  Do first run to fetch total number of rows dynamically
 
url_base = 'https://data.gov.sg/api/action/datastore_search'

# Use enter a large number in the limit parameter to retrieve all the records
parameters = {
    'resource_id' : '139a3035-e624-4f56-b63f-89ae28d4ae4c',
}

response = requests.get(url_base, params=parameters)

# View the json text returned by the API server
response_dict = response.json()
response_dict

# Traverse the dictionary to the level where the main records are located
total_number_of_rows = response_dict['result']['total']

total_number_of_rows

2183

In [42]:
#  Iterate through the API call which imits to 100 records per request.
 
resource_id = '139a3035-e624-4f56-b63f-89ae28d4ae4c'

df_carpark = []

for i in range(0, total_number_of_rows, 100):
  parameters = {
    'resource_id' : resource_id,
    'offset': i 
    }
  
  url_base = 'https://data.gov.sg/api/action/datastore_search'
  
  response = requests.get(url_base, params=parameters)

  # View the json text returned by the API server
  response_dict = response.json()
  records = response_dict['result']['records']

  df_temp = pd.json_normalize(records)

  df_carpark.append(df_temp)


In [43]:
len(df_carpark)

22

In [44]:
# Concatenate the records stored in multiple DataFrames into a single DataFrame
df_carpark = pd.concat(df_carpark, axis=0, ignore_index=True)
df_carpark

short_term_parking          car_park_type     y_coord     x_coord  \
0             WHOLE DAY      BASEMENT CAR PARK  31490.4942  30314.7936   
1             WHOLE DAY  MULTI-STOREY CAR PARK  33695.5198  33758.4143   
2             WHOLE DAY       SURFACE CAR PARK  34500.3599  29257.7203   
3               7AM-7PM       SURFACE CAR PARK  39012.6664  28185.4359   
4                    NO       SURFACE CAR PARK  38684.1754  29482.0290   
...                 ...                    ...         ...         ...   
2178          WHOLE DAY  MULTI-STOREY CAR PARK  45576.0125  29850.1522   
2179          WHOLE DAY  MULTI-STOREY CAR PARK  45166.4820  30057.2209   
2180          WHOLE DAY       SURFACE CAR PARK  45686.2734  27772.9219   
2181          WHOLE DAY  MULTI-STOREY CAR PARK  45679.7181  29935.5818   
2182          WHOLE DAY       SURFACE CAR PARK  45507.8047  28077.2305   

                 free_parking gantry_height car_park_basement night_parking  \
0                          NO          1.80                 Y           YES   
1     SUN & PH FR 7AM-10.30PM          2.10                 N           YES   
2     SUN & PH FR 7AM-10.30PM          0.00                 N           YES   
3                          NO          0.00                 N            NO   
4                          NO          0.00                 N            NO   
...                       ...           ...               ...           ...   
2178  SUN & PH FR 7AM-10.30PM          2.15                 N           YES   
2179  SUN & PH FR 7AM-10.30PM          2.15                 N           YES   
2180  SUN & PH FR 7AM-10.30PM          4.50                 N           YES   
2181  SUN & PH FR 7AM-10.30PM          2.15                 N           YES   
2182  SUN & PH FR 7AM-10.30PM          4.50                 N           YES   

                                          address car_park_decks   _id  \
0     BLK 270/271 ALBERT CENTRE BASEMENT CAR PARK              1     1   
1                       BLK 98A ALJUNIED CRESCENT              5     2   
2                             BLK 101 JALAN DUSUN              0     3   
3                  BLOCK 253 ANG MO KIO STREET 21              0     4   
4                    BLK 302/348 ANG MO KIO ST 31              0     5   
...                                           ...            ...   ...   
2178                      BLK 461 YISHUN AVENUE 6             10  2179   
2179                         BLK 468 YISHUN ST 43              4  2180   
2180     BLK 731/746 YISHUN STREET 71,72/AVENUE 5              0  2181   
2181                         BLK 478 YISHUN ST 42             11  2182   
2182                 BLK 747/752 YISHUN STREET 72              0  2183   

     car_park_no type_of_parking_system  
0            ACB     ELECTRONIC PARKING  
1            ACM     ELECTRONIC PARKING  
2            AH1     ELECTRONIC PARKING  
3           AK19         COUPON PARKING  
4           AK31         COUPON PARKING  
...          ...                    ...  
2178        Y77M     ELECTRONIC PARKING  
2179        Y78M     ELECTRONIC PARKING  
2180          Y8     ELECTRONIC PARKING  
2181        Y82M     ELECTRONIC PARKING  
2182          Y9     ELECTRONIC PARKING  

[2183 rows x 13 columns]